agent/php_stacked_segment.h

```diff
@@ -155,6 +164,386 @@
+ * nr_php_observer_fcall_begin(D) so we proceed and create stacked segment *D
+ * correctly parented as a child of *A and *D becomes the current segment.
+ *
+ * case 2 nr_php_observer_fcall_end - before a segment is ended(kept/discarded)
+ *  root <                      root                      root
+ *                               |                         |
+ *                               *A <                      *A
+ *                                                         |
+ *                                                         *B
+ *
+ * We start out with a root segment, OAPI calls nr_php_observer_fcall_begin(A),
+ * and it starts stacked segment *A and then nr_php_observer_fcall_begin(B)
+ * starts stacked segment *B as a child of *A.
+ *
+ *  root                       root
+ *   |                           |
+ *   *A                          *A
+ *   |                           |
+ *   *B                          *B <
+ *    |                           |
+ *    *C <                        c
+ *
+ * nr_php_observer_fcall_begin(C) starts segment *C as child
+ * of *B. Function C throws an uncaught exception which B does not catch so
+ * neither nr_php_observer_fcall_end(B) nor nr_php_observer_fcall_end(C) is
+ * called and *C remains the current segment. A catches the exception and
+ * nr_php_observer_fcall_end(A) is triggered. At this point we compare the
+ * current stacked_segment->metadata->This value with the execute_data->This and
+ * realize the two don't match. We check the global exception hook and see it
+ * has a value and that the global uncaught_exception_this also matches the
+ * current segment `this`. Time to apply the exception and clean up dangling
+ * segments. We pop the current segment *C and apply the exception.
+ * Because it has an exception, the segment is kept so we copy the contents of
+ * the stacked segment *C into a segment c we obtained from the slab allocator,
+ * and we make c a child of the stacked segment *B which becomes the current
+ * segment.
+ *
+ *  root                        root <
+ *   |                           |
+ *   *A <                        a
+ *    |                          |
+ *    b                          b
+ *    |                          |
+ *    c                          c
+ *
+ *
+ * But we aren't done yet.
+ * current stacked_segment->metadata->this still doesn't equal the
+ * execute_data-> this provided by nr_php_observer_fcall_end(A). We pop the
+ * current segment *B and apply the exception. Because it has an exception, the
+ * segment is kept so we copy the contents of the stacked segment *B into a
+ * segment b we obtained from the slab allocator, and we make b a child of the
+ * stacked segment *A which becomes the current segment.  Now current
+ * stacked_segment->metadata->this DOES equal the execute_data-> this
+ * provided by nr_php_observer_fcall_end(A) so it proceeds, decides to keep the
```

Comment:
Does the decision to keep segment `A` is based on `A` having kept child segment `b` (it was kept because it threw exceptions)? Or having child segment `b` is irrelevant and `A` can get discarded? If `A` will get discarded, will `b` be correctly re-parented to root segment?

Reply:
`A` can possibly get discarded.  The decision to keep/discard segments is always determined in this function: https://github.com/newrelic/newrelic-php-agent/blob/main/agent/php_execute.c#L1265 

If `A` is discarded, `b` will be reparented to root.

Reply:
The "decides to keep" was to indicate for the the example that `A` passed the checks in [here](https://github.com/newrelic/newrelic-php-agent/blob/main/agent/php_execute.c#L1265) and was kept and that segment A was chosen; otherwise, it would never be moved to the heap.